Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/checkins/cryptography/CheckInCryptography.kt

```diff
@@ -1,25 +1,135 @@
+import org.joda.time.Instant
+import javax.crypto.Mac
+import javax.crypto.spec.IvParameterSpec
+import javax.crypto.spec.SecretKeySpec
 import javax.inject.Inject
 import kotlin.random.Random
 
 class CheckInCryptography @Inject constructor(
-    @RandomStrong private val randomSource: Random,
+    @RandomStrong private val secureRandom: Random,
+    private val aesCryptography: AesCryptography
 ) {
     fun encrypt(
         checkIn: CheckIn,
         transmissionRiskLevel: Int
     ): CheckInProtectedReport {
-        // TODO
-        throw NotImplementedError()
+
+        val checkInRecord = createCheckInRecord(checkIn.checkInStart, checkIn.checkInEnd, transmissionRiskLevel)
+        val encryptionKey = getEncryptionKey(checkIn.traceLocationId.toByteArray())
+        val iv = getCryptographicSeed()
+        val encryptedCheckInRecord =
+            aesCryptography.encrypt(encryptionKey.toByteArray(), checkInRecord.toByteArray(), IvParameterSpec(iv))
+        val macKey = getMacKey(checkIn.traceLocationId.toByteArray()).toByteArray()
+        val mac = getMac(macKey, iv, encryptedCheckInRecord)
+
+        return CheckInProtectedReport.newBuilder()
+            .setEncryptedCheckInRecord(encryptedCheckInRecord.toByteString().toProtoByteString())
+            .setIv(iv.toByteString().toProtoByteString())
+            .setLocationIdHash(checkIn.traceLocationIdHash.toProtoByteString())
+            .setMac(mac.toByteString().toProtoByteString())
+            .build()
+    }
+
+    fun decrypt(
+        checkInProtectedReport: CheckInProtectedReport,
+        locationId: ByteArray
+    ): CheckInRecord {
+
+        val macKey = getMacKey(locationId).toByteArray()
+        val mac = getMac(
+            macKey,
+            checkInProtectedReport.iv.toByteArray(),
+            checkInProtectedReport.encryptedCheckInRecord.toByteArray()
+        )
+
+        if (!mac.contentEquals(checkInProtectedReport.mac.toByteArray())) throw IllegalArgumentException(
+            "Message Authentication Codes are not the same ${mac.base64()} != ${
+            checkInProtectedReport.mac.toByteArray().base64()
+            }"
+        )
+
+        val encryptionKey = getEncryptionKey(locationId)
+        val ivParameterSpec = IvParameterSpec(checkInProtectedReport.iv.toByteArray())
+        val decryptedData = aesCryptography.decrypt(
+            encryptionKey.toByteArray(),
+            checkInProtectedReport.encryptedCheckInRecord.toByteArray(),
+            ivParameterSpec
+        )
+
+        return CheckInRecord.parseFrom(decryptedData)
+    }
+
+    private fun getCryptographicSeed(): ByteArray {
+        val cryptographicSeed = ByteArray(16)
+        secureRandom.nextBytes(cryptographicSeed)
+        return cryptographicSeed
+    }
+
+    private fun createCheckInRecord(checkInStart: Instant, checkInEnd: Instant, riskLevel: Int): CheckInRecord {
+        val start = (checkInStart.seconds / 60).toInt()
+        val end = (checkInEnd.seconds / 60).toInt()
```

Comment:
Should `start` and `end` time interval be `10 min interval` ?
Pls see https://github.com/corona-warn-app/cwa-app-tech-spec/pull/87/files#diff-d50804d49c54cd0ea1d31678c893711ac77c8125fd596f5dcf65e429fb0e7b85R1372-R1387

Reply:
yes, should be 10-minute intervals.

Reply:
ok I'll fix it
This actually wasn't mentioned in the **AES-Encrypting a CheckInRecord** section
https://github.com/corona-warn-app/cwa-app-tech-spec/pull/87/files#diff-d50804d49c54cd0ea1d31678c893711ac77c8125fd596f5dcf65e429fb0e7b85R553 

also, the sample data sets 1-2 are no longer valid after the 10min adjustment
/cc @mlenkeit